### PR TITLE
Print warning when optimizations are disabled

### LINKED ISSUE #1827: Warn on startup when Bevy itself is built without optimizations
## What problem does this solve or what need does it fill?

Without optimizations Bevy is slow. Not everyone realizes that they can use `--release` to drastically improve performance. This is a frequent topic on the Discord.

## What solution would you like?

See title. This can be implemented by adding a build script to a crate like `bevy_app` that reads the `OPT_LEVEL` env var. And when the value is 0 causes a startup system to be enabled that prints a warning suggesting to pass `--release`.

## What alternative(s) have you considered?

Somehow teach Cargo to always optimize Bevy and it's dependencies.

## Additional context

https://discord.com/channels/691052431525675048/742884593551802431/828710614721363969

### PR DESCRIPTION
**Pull request by SpecificProtagonist**
# Objective
Fixes #1827

## Solution

- Check optimization level of bevy_ecs
- Print warning on startup
- ~Should the warning mention that you can enable optimizations for dependencies even in debug mode?~

### CODE REVIEW

**Comment by IceSentry** [comment]
> Should the warning mention that you can enable optimizations for dependencies even in debug mode?

This should be the default and `--release` should not be mentioned. It's nice for quickly seeing if perf is the issue, but it really shouldn't be used unless you are actually building for release. I would suggest simply linking the page on the website that explains this.

In general, I'm not sure this is required though, bevy still works fine without optimization for quickly testing stuff and adding a warning for this feels a bit too much.

File: crates/bevy_app/src/app.rs [hunk 1]
```diff
@@ -1008,6 +1010,13 @@ fn run_once(mut app: App) {
     app.update();
 }
 
+fn warn_on_missing_optimization() {
+    if bevy_ecs::OPT_LEVEL == "0" {
+        // Cannot use bevy_log, as that depends on bevy_app
+        println!("Running without optimizations. To enable them, see https://bevyengine.org/learn/book/getting-started/setup/#compile-with-performance-optimizations");
```
Comment (alice-i-cecile) [inline crates/bevy_app/src/app.rs:1016]: This should use the `warn!` macro.

File: crates/bevy_ecs/build.rs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+fn main() {
+    println!("cargo:rerun-if-changed=build.rs");
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/build.rs:2]: I don't understand how this works: why does println do anything here?

**Comment by mockersf** [comment]
Is the build script really necessary? It seems it's putting the value from env var `OPT_LEVEL` in env var `OPT_LEVEL`?

File: crates/bevy_ecs/build.rs (hunk 2 above)
Comment (mockersf) [inline crates/bevy_ecs/build.rs:2]: this would mean the script wouldn't rerun if the env var `OPT_LEVEL` changed

**Comment by mockersf** [comment]
I still dislike build scripts, but it does what it's supposed to 👍

**Comment by alice-i-cecile** [comment]
Marked as controversial and ready for final review. This is well made and does what it's supposed to, but I'm not sure that any of us are convinced it's worth the tech debt :)

**Comment by cart** [comment]
I'm going to close this out. I like this conceptually, but I don't think its worth adding build.rs into the mix.

**Pull request closed without merging**